src/engine/controls/bpmcontrol.cpp

```diff
@@ -429,25 +428,34 @@ double BpmControl::calcSyncedRate(double userTweak) {
         return rate + userTweak;
     }
 
-    const double dPrevBeat = m_pPrevBeat->get();
-    const double dNextBeat = m_pNextBeat->get();
+    const auto prevBeatPosition =
+            mixxx::audio::FramePos::fromEngineSamplePosMaybeInvalid(
+                    m_pPrevBeat->get());
+    const auto nextBeatPosition =
+            mixxx::audio::FramePos::fromEngineSamplePosMaybeInvalid(
+                    m_pNextBeat->get());
 
-    if (dPrevBeat == -1 || dNextBeat == -1) {
+    if (!prevBeatPosition.isValid() || !nextBeatPosition.isValid()) {
         m_resetSyncAdjustment = true;
         return rate + userTweak;
     }
 
-    double dBeatLength = dNextBeat - dPrevBeat;
+    const double beatLengthFrames = nextBeatPosition - prevBeatPosition;
 
     // Now that we have our beat distance we can also check how large the
     // current loop is.  If we are in a <1 beat loop, don't worry about offset.
-    const bool loop_enabled = m_pLoopEnabled->toBool();
-    const double loop_size = (m_pLoopEndPosition->get() -
-                              m_pLoopStartPosition->get()) /
-                              dBeatLength;
-    if (loop_enabled && loop_size < 1.0 && loop_size > 0) {
-        m_resetSyncAdjustment = true;
-        return rate + userTweak;
+    if (m_pLoopEnabled->toBool()) {
+        const auto loopStartPosition =
+                mixxx::audio::FramePos::fromEngineSamplePosMaybeInvalid(
+                        m_pLoopStartPosition->get());
+        const auto loopEndPosition =
+                mixxx::audio::FramePos::fromEngineSamplePosMaybeInvalid(
+                        m_pLoopEndPosition->get());
+        const auto loopSize = (loopEndPosition - loopStartPosition) / beatLengthFrames;
```

Comment:
Wouldn't this trigger a debug assertion if loopStartPosition or loopEndPosition is invalid? Using fromEngineSamplePosMaybeInvalid() doesn't help to prevent this. Either we are sure that both positions are valid at this point or otherwise we should skip the code entirely.

Reply:
Yes, it would trigger a debug assertion. However, in that case the loop enabled CO should be false. But I guess there could be a race condition if the CO changes in between. I'll add a check.